azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSLocation.java

```diff
@@ -50,27 +59,23 @@ class ADLSLocation {
     Preconditions.checkArgument(location != null, "Invalid location: null");
 
     Matcher matcher = URI_PATTERN.matcher(location);
-
-    ValidationException.check(matcher.matches(), "Invalid ADLS URI: %s", location);
-
-    String authority = matcher.group(1);
-    String[] parts = authority.split("@", -1);
-    if (parts.length > 1) {
-      this.container = parts[0];
-      this.storageAccount = parts[1];
-    } else {
-      this.container = null;
-      this.storageAccount = authority;
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException(String.format("Invalid ADLS URI: %s", location));
     }
 
-    String uriPath = matcher.group(2);
-    uriPath = uriPath == null ? "" : uriPath.startsWith("/") ? uriPath.substring(1) : uriPath;
-    this.path = uriPath.split("\\?", -1)[0].split("#", -1)[0];
+    try {
+      URI uri = new URI(location);
```

Comment:
Not using `java.net.URI` before was intentional, in the past we have avoided using `java.net.URI` for parsing as it can cause some unexpected results. For example, underscores in host names...

```
jshell> new java.net.URI("abfs://my_endpoint/path").getHost()
$1 ==> null
```

(cc @rdblue since you've pointed this out to me)

Reply:
I think that's fine, but if this is the case we need to add in some explicit tests for this scenario. @mrcnc would you like to add a follow-on pr?

Reply:
I believe we will not encounter this error b/c [underscores aren't valid characters for storage accounts or container names](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftstorage).   But maybe we should add more input validation or update the regex before attempting to use use `java.net.URI`?  I was thinking using URI would be more idiomatic but if we have reasons to avoid using it then we can revert or refactor

Reply:
I wish we had caught this before now, but I think we need to revert the URI additions.  There are a number of issues related to the Java implementation which is why we specifically avoid using it.

Reply:
Could we have some examples of where this doesn't work?